modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,101 +5,167 @@
 
-export type FormatCallbacks = Record<string, FormatChangeCallback[]>;
+export type RegisteredFormats = Record<string, CallbackGroup>;
 
-export interface FormatData {
-  similar?: boolean;
-  callbacks: FormatChangeCallback[];
+export interface UnbindFormatChanged {
+  unbind: () => void;
 }
 
-export type RegisteredFormats = Record<string, FormatData>;
+interface CustomCallback {
+  readonly state: Cell<boolean>;
+  readonly similar: boolean;
+  readonly vars: FormatVars;
+  readonly callback: FormatChangeCallback;
+}
 
-export interface UnbindFormatChanged {
-  unbind: () => void;
+interface CallbackGroup {
+  // Bundle all callbacks with similar=true and unspecified vars
+  readonly similarState: Cell<boolean>;
+  readonly similarCallbacks: FormatChangeCallback[];
+  // Also bundle all callbacks with similar=false and unspecified vars
+  readonly similarFalseState: Cell<boolean>;
+  readonly similarFalseCallbacks: FormatChangeCallback[];
+  // Then for callbacks that have variables, handle everything separately
+  readonly others: Array<CustomCallback>;
 }
 
 const setup = (registeredFormatListeners: Cell<RegisteredFormats>, editor: Editor) => {
-  const currentFormats = Cell<Record<string, FormatChangeCallback[]>>({ });
-
   registeredFormatListeners.set({});
 
   editor.on('NodeChange', (e) => {
-    updateAndFireChangeCallbacks(editor, e.element, currentFormats, registeredFormatListeners.get());
+    updateAndFireChangeCallbacks(editor, e.element, registeredFormatListeners.get());
   });
 };
 
-const updateAndFireChangeCallbacks = (editor: Editor, elm: Element, currentFormats: Cell<FormatCallbacks>, formatChangeData: RegisteredFormats) => {
-  const formatsList = Obj.keys(currentFormats.get());
-  const newFormats: FormatCallbacks = { };
-  const matchedFormats: FormatCallbacks = { };
+const matchingNode = (editor: Editor, parents: Element[], format: string, similar: boolean, vars?: FormatVars): Optional<Element> =>
+  Arr.findUntil(
+    parents,
+    (node) => { // is this the node we want?
+      const matchingFormat = editor.formatter.matchNode(node, format, vars ?? {}, similar);
+      return !Type.isUndefined(matchingFormat);
+    },
+    (node) => { // Can we guarantee that no parent of this will be the node we want?
+      if (MatchFormat.matchesUnInheritedFormatSelector(editor, node, format)) {
+        return true;
+      } else {
+        if (!similar) {
+          // If we want to find an exact match, then finding a similar match halfway up the parents tree is bad
+          return Type.isNonNullable(editor.formatter.matchNode(node, format, vars, true));
+        } else {
+          return false;
+        }
+      }
+    }
+  );
+
+const getParents = (editor: Editor, elm?: Element): Element[] => {
+  const element = elm ?? editor.selection.getNode();
+  return Arr.filter(FormatUtils.getParents(editor.dom, element), (node) =>
+    node.nodeType === 1 && !node.getAttribute('data-mce-bogus')
```

Comment:
I know this was like this previously, but since you've done a lot of refactoring/cleaning up here, can we please use the `NodeType.isElement` function here instead of the `nodeType` comparision?